Reference: Add suggest API guide and custom index settings docs (#687)
inveniosoftware/docs-invenio-rdm: docs/reference/search.md

## How to use suggest API?

InvenioRDM's Suggest API is designed to provide search suggestions by using a `multi_match` query. It can be configured for all the indices using the `SuggestQueryParser` class that can be imported from `invenio-records-resources` module. The fields are analyzed using custom analyzers at index time which apply filters like `asciifolding` for accent search and `edge_ngram` to generate tokens of all lengths according to the filter settings to power `prefix matching` search capabilities in the dropdown UI fields.

Check the [official documentation](https://opensearch.org/docs/2.0/opensearch/ux/) and [Reference](#reference) for more context on the `edge_ngram` filter and custom analyzers.

### When to Use the Suggest API

- **Typo Tolerance & Auto-completion:** Helps correct typos (using `fuzziness` at search time analyzing) and completes partial inputs.
- **Large, Diverse Datasets:** Useful for datasets with a wide variety of terms, like names or titles.
- **Pre-query Optimization:** Reduces unnecessary searches by suggesting relevant terms.

### When Not to Use the Suggest API

- **Small or Specific Datasets:** Less beneficial for well-defined datasets.

> - **Performance Constraints:** Because the suggest API creates large amounts of token using the `edge_ngram` filter, it is important to observe how it affects the index size.

Reviewer: ```suggestion
- **Performance Constraints:** Because the suggest API creates large amounts of tokens using the `edge_ngram` filter, it is important to observe how it affects the index size.
```

  - A reasonable trade-off might involve an index size increase of up to 20-30% if it significantly improves search speed and relevance.
  - A 10-20% improvement in response times might justify a moderate increase in index size.

For more information check the [official documentation](https://www.elastic.co/guide/en/elasticsearch/reference/current/tune-for-disk-usage.html).

## Key Considerations for Customizing Index Mappings

### Size

- Field Type Selection: Use lightweight field types (e.g., keyword over text where appropriate) to minimize index size.
- Usage of custom analyzers or filters should be as limited as possible to prevent index bloating.

### Speed

> - Search Performance: Kepeping size in mind, apply custom analyzers that include `edge_ngram` filter for providing quick suggestions, and optimize for frequently queried fields to enhance search speed.

Reviewer: ```suggestion
- Search Performance: Keeping size in mind, apply custom analyzers that include `edge_ngram` filter to provide quick suggestions, and optimize for frequently queried fields to enhance search speed.
```

- Analyzer and filter selection: Configure only when necessary to improve search time.

## Fine tuning the search

> - **Relevance Adjustment:** Boosting affects the relevance score of documents. A higher boost value means a stronger influence on the search ranking. Determine which fields are most critical for your search relevance (e.g., titles, authors, keywords).
> 
> Boosting of field(s) can be done by using the caret operator **(^)** followed by a number.
> Examples: `name^100` will boost the name field by a factor of 100.
> Asterisk **(\*)** can be used to apply boosting to all the subfields. `i18n_titles.*^50`
> 
> - **Balance and Tuning:** Use boosting judiciously to avoid skewing results too heavily towards particular fields. Assign boost factors based on the importance of each field. Higher values increase the influence of matches in that field.

Reviewer: ```suggestion
Boosting affects the relevance score of documents. A higher boost value means a stronger influence on the search ranking. Determine which fields are most critical for your search relevance (e.g., titles, authors, keywords).

- **Relevance Adjustment:** Boosting of field(s) can be done by using the caret operator **(^)** followed by a number. For example:
    * `name^100` will boost the name field by a factor of 100.
    * Asterisk **(\*)** can be used to apply boosting to all the subfields. `i18n_titles.*^50`

- **Balance and Tuning:** Use boosting judiciously to avoid skewing results too heavily towards particular fields. Assign boost factors based on the importance of each field. Higher values increase the influence of matches in that field.
```

I think this works better having an introductory sentence first

# Searching in InvenioRDM

### InvenioRDM Suggest API

> The suggest API endpoint (`/api/{resource}/?suggest={search_input}`) provides an interface for real-time search suggestions. It leverages OpenSearch's `multi_match` query to search across multiple fields within a specified index, returning relevant suggestions based on user input.

Reviewer 2: ```suggestion
The suggest API endpoint (`/api/{resource}?suggest={search_input}`) provides an interface for real-time search suggestions. It leverages OpenSearch's `multi_match` query to search across multiple fields within a specified index, returning relevant suggestions based on user input.
```

> ### Endpoint Structure

Reviewer 2: "Endpoint Structure" could be moved inside the "InvenioRDM Suggest API" chapter

> **URL:** `/api/{index-name}?suggest={search_input}`

Reviewer 2: Is there a difference between `index-name` and `resource`? `index-name` sounds like the internal index name (e.g. `zenodo-qa-names-name-v1.0.0-1708098159`). We should be consistent with the terminology use.

- **Large, Diverse Datasets:** Useful for datasets with a wide variety of terms, like names or titles.
- **Pre-query Optimization:** Reduces unnecessary searches by suggesting relevant terms.

### When Not to Use the Suggest API

- **Small or Specific Datasets:** Less beneficial for well-defined datasets.
- **Performance Constraints:** Because the suggest API creates extra subfields like `._2gram`, `._3gram` and `._index_prefix`, it is important to observe how it affects the index size.
  - A reasonable trade-off might involve an index size increase of up to 20-30% if it significantly improves search speed and relevance.
  - A 10-20% improvement in response times might justify a moderate increase in index size.

For more information check the [official documentation](https://www.elastic.co/guide/en/elasticsearch/reference/current/tune-for-disk-usage.html).

## Key Considerations for Customizing Index Mappings

### Size

- Field Type Selection: Use lightweight field types (e.g., keyword over text where appropriate) to minimize index size.
- Usage of custom analyzers or filters should be as limited as possible to prevent index bloating.

### Speed

> - Search Performance: Implement search_as_you_type or completion fields for quick suggestions, and optimize for frequently queried fields to enhance search speed.

Reviewer 2: ```suggestion
- Search Performance: Implement `search_as_you_type` or `completion` fields for quick suggestions, and optimize for frequently queried fields to enhance search speed.
```

Analyzers can be applied to both the search input and when the document is indexed. In most cases we want to apply the same analyzer to the search input and during indexing so that there is not unexpected behaviour.

- [**Normalizers**](https://opensearch.org/docs/latest/analyzers/normalizers/) — Simpler and mainly used to improve the matching of keyword search. The `keyword` type is the simplest way in which data can be stored and by default works as an exact match search. Using a normalizer you can add, remove and alter the input into exactly one other token which is stored and searched for.

### Character filters

Character filters take the stream of characters before tokenization and can add, remove or replace characters according to the rules and type of filter. There are 3 types - mapping filter, pattern replace filter and HTML stripping filter.

For our indices in InvenioRDM, we are currently using a custom pattern replace filter that uses regex to remove special characters!

```
"char_filter": {
  "strip_special_chars": {
    "type": "pattern_replace",
    "pattern": "[\\p{Punct}\\p{S}]",
    "replacement": ""
  }
}
```

### Tokenizers and token filters

> 

Reviewer 2: ```suggestion

We are using the following tokenizers and token filters in some of our indices in InvenioRDM:

```
